01.Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

02.Importing Data

In [2]:
path = r'/Users/faridchehraz/Desktop/Python/Master - Dec2021 Instacart Basket Analysis'

In [3]:
df_prods = pd.read_csv(os.path.join(path, '2. Data', 'a. Original Data', 'products.csv'), index_col = False)

In [5]:
df_ords = pd.read_csv(os.path.join(path, '2. Data', 'b. Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
# Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a','b', '1', True]

In [8]:
df_test.head

<bound method NDFrame.head of     mix
0     a
1     b
2     1
3  True>

In [10]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [11]:
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

In [13]:
df_prods.isnull( ).sum( )

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [22]:
df_prods_clean_no_dups.shape

(49672, 5)

In [23]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'2. Data', 'b. Prepared Data', 'products_checked.csv'))

In [24]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#Prices column: Max value of 99,999 is worth investigating. Unlikely an online grocery store would sell
#an item for 99k

#Product_id column: Count is > max value by 5. Could be the duplicates prevously identified

In [28]:
#Check for mixed data types in df_ords dataframe - none found
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [29]:
#Check for missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [33]:
#Create subset of ords dataframe with null values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [34]:
#View subset of ords dataframe with null values
df_nan_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

#The only missing values in the ords dataframe are in the days_since_prior columns.

#Looking at some of the output and thinking it through logically, I would expect records to be blank for

#first time customers i.e. there would be nothing in this column for first time customers.

#For this reason I would not attempt to impute values or delete records.

#This could be something however to discuss with the client.

In [37]:
#Check for duplicate values in df_ords dataframe - no duplicates found
df_dups_ords = df_ords[df_ords.duplicated()]

In [38]:
df_dups_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [39]:
df_ords.to_csv(os.path.join(path,'2. Data', 'b. Prepared Data', 'orders_checked.csv'))